### PR TITLE
Create security_baseline.md

### PR DESCRIPTION
**Pull request by Danajoyluck**
GitHub version of the security baseline for TAC review.

### CODE REVIEW

File: process/baseline/security_baseline.MD
Comment (lehors) [inline process/baseline/security_baseline.MD:15]:
```suggestion
In the United States, open source software is used across all critical infrastructure sectors defined by CISA (Cybersecurity and Infrastructure Security Agency), for example, health care, defense, financial services, utilities, telecommunications, etc. Open source security directly impacts national security, economics and social stability. Enhancing open source security is imperative. NIST has published a [Secure Software Development Framework](https://csrc.nist.gov/Projects/ssdf) (SSDF) as a result of Executive Order (EO) 14028 on "[Improving the Nation's Cybersecurity](https://www.federalregister.gov/documents/2021/05/17/2021-10460/improving-the-nations-cybersecurity)".
```

File: process/baseline/security_baseline.MD
Comment (lehors) [inline process/baseline/security_baseline.MD:235]:
```suggestion
### Curl and OpenSSL Tips
```

File: process/baseline/security_baseline.MD
Comment (lehors) [inline process/baseline/security_baseline.MD:72]:
```suggestion
      * Incorporate the baseline into OpenSSF Technical Advisory Council (TAC) [technical initiative life cycle process](https://github.com/ossf/tac/blob/main/process/project-lifecycle.md).
```

**Comment by lehors** [comment]
LGTM but I have a couple of nits and a comment/question:
- I don't think we should add a `baseline` folder, just add the file in the `process` folder.
- the file extension should be `.md` rather than `.MD`

If, as discussed, we create a SIG to work on this, are we expecting them to work on this document in the TAC repo? That doesn't seem really desirable. Maybe they could create a copy of this document to work on in another repo. The instance in the TAC repo would then be refreshed from time to time when the SIG produces a new version that the TAC agrees to adopt. So, the instance in the TAC repo would remain the governing version while another copy exists in some other repo for the SIG to work on. Does that sound like a workable process? Any other idea?

**Comment by mlieberman85** [comment]
> LGTM but I have a couple of nits and a comment/question:
> 
> * I don't think we should add a `baseline` folder, just add the file in the `process` folder.
> * the file extension should be `.md` rather than `.MD`
> 
> If, as discussed, we create a SIG to work on this, are we expecting them to work on this document in the TAC repo? That doesn't seem really desirable. Maybe they could create a copy of this document to work on in another repo. The instance in the TAC repo would then be refreshed from time to time when the SIG produces a new version that the TAC agrees to adopt. So, the instance in the TAC repo would remain the governing version while another copy exists in some other repo for the SIG to work on. Does that sound like a workable process? Any other idea?

I agree that we should have something external and only when read do we bring it into the TAC repo. I am making the assumption here that any material change to the baseline would require the TAC to approve.

**Comment by lehors** [comment]
> I am making the assumption here that any material change to the baseline would require the TAC to approve.

Exactly. That's my assumption too. I don't think we want to delegate the responsibility of managing the baseline for the whole organization to the SIG.

**Comment by david-a-wheeler** [comment]
Is this baseline *only* for the OpenSSF, or is the intention for this be a baseline applicable to *many* foundations? Perhaps at one time this was OpenSSF-only, but I think the longer goal is to create a "security baseline" that's useful for many foundations.

It's *fine* to put in the TAC repo in the short term, but I think in the long term it should *NOT* be in the TAC repo. It should instead eventually have its own repo or be part of the Best Practices WG repo (the latter would probably be easiest). I think that location will provide a stronger indication that this baseline can be used by other foundations.

**Comment by Danajoyluck** [comment]
> Is this baseline _only_ for the OpenSSF, or is the intention for this be a baseline applicable to _many_ foundations? Perhaps at one time this was OpenSSF-only, but I think the longer goal is to create a "security baseline" that's useful for many foundations.
> 
> It's _fine_ to put in the TAC repo in the short term, but I think in the long term it should _NOT_ be in the TAC repo. It should instead eventually have its own repo or be part of the Best Practices WG repo (the latter would probably be easiest). I think that location will provide a stronger indication that this baseline can be used by other foundations.

The baseline is for OpenSSF. Can we put it here for now so that we can start updating the life cycle document and get the adoption going?

**Comment by david-a-wheeler** [comment]
@lehors said:

> My take is that this is only meant for OpenSSF for now. When a SIG is launched with a goal to create one for the larger community they can create their own. Then, the TAC can decide how/when to revise this one.

Oh, I see. *This* PR is for an OpenSSF-specific baseline, and there will likely be a SIG created that starts with & builds on this material to create something for multiple foundations. Presumably, when that SIG is done, the TAC will decide if this PR's text will be replaced. That makes sense. Thanks so much for the clarification!

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:17]: I don't know if this is how the CRA text is written, but the term "manufacturer" isn't typically used in the context of software. In SLSA, for example, the term we have agreed upon is the "software producer", so I would recommend using a term like "software producer" whenever referring to the entity that creates a software product.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:20]: nit
```suggestion
The success of the security baseline SHOULD be quantified and qualified in a few dimensions:
```

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:21]: This might just be my nitpicking, but it wasn't clear to me what you meant by "security outcomes" at first. To me, that sounds like this is referring to the improvement of the overall security posture, but this success criterion is really about adopting and advancing tooling/frameworks/etc. I suggest renaming this criterion to something that more clearly conveys the object and approach:

```suggestion
 * **Adoption and Advancement of Security Technologies**
```

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:25]: As for the others, are there specific metrics that could be used as measures of success here?

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:26]: Just to make the distinction with the previous criterion clear, I suggest a slight phrasing change here as well:
```suggestion
 * **Increased Security Baseline Adoption Rate**
```

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:31]: What do we mean by "adoption" here? Is it enough for the other foundations to agree to the adoption, or is the expectation that they would have a foundation-wide implementation of the security baseline by the end of 2024? The former seems doable by the end of the year, the latter would be more challenging.

Also, I believe it was mentioned this one was a stretch goal for 2024. I would label it as such.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:37]: Similar to my comment above, can we come up with a metric or two that would allow us to measure that maintainer/consumer experience has improved? I know this is incredibly hard, but finding some key measurable results may help in addressing future gaps etc.

Comment (marcelamelara) [reply process/security_baseline.md:37]: On the other hand, if we don't intend to measure this, I can see this point fitting under operating principles as well.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:57]: This makes sense as an objective, but the approach, i.e. how we plan/recommend to prioritize and achieve automation/automatability, is not. Is there a more concrete approach we could describe here? Are there existing best practices this document could refer to?

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:63]: I'm a little unclear as to what these two approaches really mean. 
* How does shift-left lead to incremental adoption? 
* With "Reuse existing guides/technologies without imposing new requirements", is the idea here that these guides won't change significantly adding more requirements? The reason I'm also a little confused here is that the security baseline itself _will_ likely impose new requirements on projects, so clarifying what "new requirements" are meant here would be helpful.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:64]: Is this principle about continuous improvement of the security baseline? This was not immediately clear in the Objective and Approach.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:70]: This objective sounds a like the the previous operating principle. I would rephrase this to make the governance-related objective clearer.

**Review by marcelamelara — CHANGES_REQUESTED** [review]
Thanks for submitting this PR @Danajoyluck ! I've reviewed the first half of the document, and I like the direction this is going. I do have several comments around clarity and specificity that will hopefully help improve this document.

File: process/security_baseline.md
Comment (Danajoyluck) [reply process/security_baseline.md:17]: > I don't know if this is how the CRA text is written, but the term "manufacturer" isn't typically used in the context of software. In SLSA, for example, the term we have agreed upon is the "software producer", so I would recommend using a term like "software producer" whenever referring to the entity that creates a software product.

Thanks @marcelamelara I will use software producer in other places. In CRA context, manufacturer needs to be used. I had to update my presentation to replace producer with manufacturer when talking about OpenSSF in light of CRA

File: process/security_baseline.md
Comment (Danajoyluck) [reply process/security_baseline.md:37]: @marcelamelara can we solicit feedback from maintainers during adoption about experience improvement measure? will a survey by the end of the adoption piloting be helpful?

File: process/security_baseline.md
Comment (Danajoyluck) [reply process/security_baseline.md:57]: @marcelamelara we will start tracking the adoption once it starts, through GitHub issues. The tracking includes adoption friction points, additional tooling needs, gaps in the baseline. This is echoing the same sentiment as security tool belt, how we can make security adoption easier through tooling.

File: process/security_baseline.md
Comment (Danajoyluck) [reply process/security_baseline.md:63]: @marcelamelara 
How does shift-left lead to incremental adoption? we embed security into the SDLC process, that's shifting security to the left. At various life cycle, the security risks of the projects are different, only the necessary security measures are taken into account for that life cycle, and build it up gradually as the software becomes more mature, that's the incremental approach supported by the shifting security to the left. What is a better way to represent this? 

Around "Reuse existing guides/technologies without imposing new requirements", the baseline aggregates the best security practices and technologies that are essential for developing secure software. I think we need to have a discussion about how we evolve the guides and the security baseline in harmony. How about we rephrase it to "Reuse existing guides/technologies with minimal new requirements"?

File: process/security_baseline.md
Comment (Danajoyluck) [reply process/security_baseline.md:64]: the three bullets under "Minimal, Achievable, and Practical Baseline Requirements" meant to ensure that the baseline is not adding additional burdens to the maintainers while adopting the baseline. The continuous improvement is more focused on  we reduce security burden on maintainers through continuous improvements,

File: process/security_baseline.md
Comment (lumjjb) [inline process/security_baseline.md:111]: We had an issue with this in this during the prometheus security review in CNCF. I think the question here is depending on the model... E.g. if you run a service on istio, you would have TLS, but you wouldn't want the servers to actually run TLS themselves too... This is a weird one because its subjective on deployment and threat model. I would suggest against making this a requirement.

File: process/security_baseline.md
Comment (lumjjb) [inline process/security_baseline.md:109]: Would this include publishing of VEX statements and ensuring that scorecard is able to interpret the VEX formats as well?

**Review by sevansdell — APPROVED** [review]
Thanks for requested updates.

File: process/security_baseline.md
Comment (marcelamelara) [reply process/security_baseline.md:17]: Interesting, thanks for clarifying.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:97]: These table entries are quite difficult to read because of all of the in-lined links. Can you please put all URLs at the bottom of the document to make this table a bit clearer?

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:82]: I think it would be helpful to add a short summary of the overall goal of the Once Sandbox baseline, i.e. how do memory-safe languages and 2FA help achieve that goal?

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:90]: This says "projects SHOULD onboard Scorecard", but it seems like there are one or two required Scorecard metrics below. [Current TI incubating requirements](https://github.com/ossf/tac/blob/main/process/TI-Gives%2BGets.md#givesrequirements-1) already include Scorecard as a MUST, so I think the baseline ought to be consistent with that and also state Scorecard onboarding as a MUST. The table below can then clarify which specific Scorecard metrics are needed for incubating.

File: process/security_baseline.md
Comment (marcelamelara) [inline process/security_baseline.md:101]: As above, what's the overall security goal of this baseline level?

File: process/security_baseline.md
Comment (marcelamelara) [reply process/security_baseline.md:110]: +1

**Review by marcelamelara — COMMENTED** [review]
Thanks for the updates @Danajoyluck ! I left some specific comments, but have two high-level asks: 1) Can you please move the in-lined links in the security baseline tables to the bottom of the document? The actual text in the tables is currently quite hard to read and review. 
2) To help motivate adoption and provide clear reasoning for each baseline level, I suggest stating the overall security goal for each level. For example, at "Once Sandbox" the overall goal is to avoid some of the most common security issues from the start. This will also help explain why each requirement was chosen for that particular level.

**Comment by Danajoyluck** [comment]
> Thanks for the updates @Danajoyluck ! I left some specific comments, but have two high-level asks: 1) Can you please move the in-lined links in the security baseline tables to the bottom of the document? The actual text in the tables is currently quite hard to read and review. 2) To help motivate adoption and provide clear reasoning for each baseline level, I suggest stating the overall security goal for each level. For example, at "Once Sandbox" the overall goal is to avoid some of the most common security issues from the start. This will also help explain why each requirement was chosen for that particular level.

Thanks a lot @marcelamelara for taking the time.

It will be very hard for maintainers to read the document if the links are not inline. Will this file view help with the review? https://github.com/ossf/tac/blob/22639387114e387c5f3f303c0d30ad88e9a79c1e/process/security_baseline.md

Around adoption, at least three projects have agreed to adopt. can we take an agile approach on this? I put the reasons why we need the baseline for each "level", would you consider that as goals?

**Review by SecurityCRob — APPROVED** [review]
I approve

File: process/security_baseline.md
Comment (puerco) [inline process/security_baseline.md:29]: This sounds reasonable, as mentioned on https://github.com/ossf/tac/issues/215 both @protobom and @openvex will start piloting the baseline as soon as it is merged. Especially protobom is still working on setting its new org, so its an ideal time for us.

**Review by mlieberman85 — APPROVED** [review]
LGTM, there's a few caveats but as mentioned in chat I think we are at the point where we can iterate in the SIG and among the pilot projects like GUAC, OpenVEX, protobom, etc.

**Review by SecurityCRob — APPROVED** [review]
I look forward to seeing how our pilot projects work through achieving the baseline!

**Review by lehors — APPROVED** [review]
LGTM

**Review by torgo — APPROVED** [review]
lgtm

**Review by sevansdell — APPROVED** [review]

**Review by bobcallaway — APPROVED** [review]

**Review by SecurityCRob — APPROVED** [review]

File: process/security_baseline.md
Comment (steiza) [inline process/security_baseline.md:124]:
```suggestion
|If release artifacts are created, SBOM is generated and distributed as a release artifact.|To provide accurate and complete inventory of the software version, license,  dependencies and vulnerabilities to enhance supply chain security, in both human and  machine processable format.|Refer to [OpenSSF SBOM Everywhere SIG](https://github.com/ossf/sbom-everywhere) for the importance of  SBOM generation, consumption and process improvement.<br /><br />Generate SBOM’s in your project’s GitHub release workflow, ensure [minimal data fields recommended by NTIA](https://www.ntia.gov/report/2021/minimum-elements-software-bill-materials-sbom) are populated.<br /><br />Example SBOM: <br />OpenSSF: [GUAC](https://github.com/guacsec/guac/blob/main/.github/workflows/release.yaml) SPDX<br />CNCF: [Kyverno](https://github.com/kyverno/kyverno/blob/68df5af40e9820633162e8d1c0b9966032eb7eb8/.github/actions/publish-image/action.yaml#L3) CycloneDX<br /><br />SECURITY_INSIGHTS.yml SHALL be updated under “dependencies” > “sbom” to include all the build SBOM’s.<br /><br />[TAC project lifecycle governance process](https://github.com/ossf/tac/blob/main/process/project-lifecycle.md) SHALL be used for exception approval when there is no tooling support to implement the baseline.|SECURITY_INSIGHTS.yml identifies the SBOM’s. SMOB’s provide information for dependency management, vulnerability management. |
```

**Review by mlieberman85 — APPROVED** [review]
LGTM

File: process/security_baseline.md
Comment (steiza) [inline process/security_baseline.md:126]: Minor: consider SLSA Build v1.0 Level 2+

In working with people to adopt SLSA Build Level 3, it's not always clear what that entails. My interpretation is that when you **verify** the provenance, you are checking that the workflow matches one you've vetted / audited. Of course, the project **producing** the provenance won't necessarily control how other people verify it. It's not (in my opinion) as simple as saying "use reusable workflows", as you don't actually get the SLSA Build Level 3 benefit unless you're verifying which workflow - reusable or not, at a specific revision - was used in the build.

File: process/security_baseline.md
Comment (steiza) [inline process/security_baseline.md:131]:
```suggestion
### Security Baseline - Once Graduated
```

**Review by steiza — APPROVED** [review]
I have a bunch of minor, nit-picky feedback. Great job synthesizing this complex topic, and I can't wait to get feedback from projects as we roll this out!

**Review by torgo — APPROVED** [review]
lgtm

**Review by marcelamelara — APPROVED** [review]
Thanks for the updates @Danajoyluck ! There are still some edits I'd like to see eventually, but I think this is ready for us and the SIG to begin iterating over.

**Pull request merged**
